cpu/lpc11u34/periph/gpio.c

```diff
@@ -0,0 +1,501 @@
+    res = gpio_init(dev, GPIO_DIR_IN, pullup);
+    if (res < 0) {
+        return res;
+    }
+
+    /* Disable interrupts */
+    NVIC_DisableIRQ(gpio_int_id);
+    /* Register pin as flex interrupt */
+    LPC_SYSCON->PINTSEL[gpio_int_id] = (port * 24) + pin;
+
+    /* set callback */
+    gpio_config[gpio_int_id].cb = cb;
+    gpio_config[gpio_int_id].arg = arg;
+
+    /* configure the event that triggers an interrupt */
+    switch (flank) {
+        case GPIO_RISING:
+            LPC_GPIO_PIN_INT->ISEL &= ~(1 << gpio_int_id); /* Set PMODE=edge sensitive */
+            LPC_GPIO_PIN_INT->IENR |= (1 << gpio_int_id);  /* Enable rising edge. */
+            LPC_GPIO_PIN_INT->IENF &= ~(1 << gpio_int_id); /* Disable falling edge. */
+            break;
+        case GPIO_FALLING:
+            LPC_GPIO_PIN_INT->ISEL &= ~(1 << gpio_int_id); /* Set PMODE=edge sensitive */
+            LPC_GPIO_PIN_INT->IENR &= ~(1 << gpio_int_id); /* Disable rising edge. */
+            LPC_GPIO_PIN_INT->IENF |= (1 << gpio_int_id);  /* Enable falling edge. */
+            break;
+        case GPIO_BOTH:
+            LPC_GPIO_PIN_INT->ISEL &= ~(1 << gpio_int_id); /* Set PMODE=edge sensitive */
+            LPC_GPIO_PIN_INT->IENR |= (1 << gpio_int_id);  /* Enable rising edge. */
+            LPC_GPIO_PIN_INT->IENF |= (1 << gpio_int_id);  /* Enable falling edge. */
+            break;
+    }
+
+    /* clear any pending requests */
+    LPC_GPIO_PIN_INT->RISE = (1 << gpio_int_id); /* Clear rising edge (sort of) flag */
+    LPC_GPIO_PIN_INT->FALL = (1 << gpio_int_id); /* Clear falling edge (sort of) flag */
+
+    NVIC_SetPriority(gpio_int_id, 3);
+
+    flex_int_mapping[dev] = gpio_int_id+1;
+
+    gpio_int_id++;
+
+    return 0;
+}
+
+void gpio_irq_enable(gpio_t dev)
+{
+    if (dev >= GPIO_NUMOF) {
+        return;
+    }
+
+    NVIC_EnableIRQ(flex_int_mapping[dev]-1);
+}
+
+void gpio_irq_disable(gpio_t dev)
+{
+    uint8_t int_id;
+
+    if (dev >= GPIO_NUMOF) {
+        return;
+    }
+
+    int_id = flex_int_mapping[dev] - 1;
+
+    NVIC_DisableIRQ(int_id);
+}
+
+int gpio_read(gpio_t dev)
+{
+    uint8_t port;
+    uint8_t pin;
+
+    if (dev >= GPIO_NUMOF) {
+        return -1;
+    }
+
+    port = gpio_port_map[dev];
+    pin = gpio_pin_map[dev];
+
+    if(LPC_GPIO->PIN[port] & (1 << pin)) /* read output data register */
+        return 1;
+    else
+        return 0;
+}
+
+void gpio_set(gpio_t dev)
+{
+    uint8_t port;
+    uint8_t pin;
+
+    if (dev >= GPIO_NUMOF) {
+        return;
+    }
+
+    port = gpio_port_map[dev];
+    pin = gpio_pin_map[dev];
+
+    LPC_GPIO->SET[port] = (1 << pin);
+}
+
+void gpio_clear(gpio_t dev)
+{
+    uint8_t port;
+    uint8_t pin;
+
+    if (dev >= GPIO_NUMOF) {
+        return;
+    }
+
+    port = gpio_port_map[dev];
+    pin = gpio_pin_map[dev];
+
+    LPC_GPIO->CLR[port] = (1 << pin);
+}
+
+void gpio_toggle(gpio_t dev)
+{
+    if (gpio_read(dev)) {
+        gpio_clear(dev);
+    } else {
+        gpio_set(dev);
+    }
```

Comment:
I know we are getting into optimisation already, but couldn't this be done by
```C
void gpio_toggle(gpio_t dev)
{
    if (dev >= GPIO_NUMOF) {
        return;
    }
    LPC_GPIO->NOT[gpio_port_map[dev]] = (1 << gpio_pin_map[dev]);
}
```
Saves us one if-statement and two function calls.